cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,329 @@
+    gdf_size_type const num_destination_rows,
+    gdf_size_type const* const __restrict__ merged_table_indices,
+    gdf_size_type const* const __restrict__ merged_row_indices) {
+
+  gdf_index_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+
+  auto active_threads =
+      __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  while (destination_row < num_destination_rows) {
+    bit_mask_t const* source_mask = merged_table_indices[destination_row] == LEFT_TABLE ? source_left_mask : source_right_mask;
+    bool const source_bit_is_valid{
+        source_mask ? bit_mask::is_valid(source_mask, merged_row_indices[destination_row]) : true};
+    
+    // Use ballot to find all valid bits in this warp and create the output
+    // bitmask element
+    bit_mask_t const result_mask{
+        __ballot_sync(active_threads, source_bit_is_valid)};
+
+    gdf_index_type const output_element = cudf::util::detail::bit_container_index<bit_mask_t, gdf_index_type>(destination_row);
+    
+    // Only one thread writes output
+    if (0 == threadIdx.x % warpSize) {
+      destination_mask[output_element] = result_mask;
+    }
+
+    destination_row += blockDim.x * gridDim.x;
+    active_threads =
+        __ballot_sync(active_threads, destination_row < num_destination_rows);
+  }
+}
+
+void materialize_bitmask(gdf_column const* left_col,
+                        gdf_column const* right_col,
+                        gdf_column* out_col,
+                        gdf_size_type const* table_indices,
+                        gdf_size_type const* row_indices,
+                        cudaStream_t stream) {
+    constexpr gdf_size_type BLOCK_SIZE{256};
+    cudf::util::cuda::grid_config_1d grid_config {out_col->size, BLOCK_SIZE };
+    // TODO: cant use just one template flag for valids, so need to test for null explicitly
+    // or make the kernel receive just one input col and call it twice one for each input column
+    materialize_merged_bitmask_kernel
+    <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+    (reinterpret_cast<bit_mask_t*>(left_col->valid),
+     reinterpret_cast<bit_mask_t*>(right_col->valid),
+     reinterpret_cast<bit_mask_t*>(out_col->valid),
+     out_col->size,
+     table_indices,
+     row_indices);
+
+    CHECK_STREAM(stream);
+}
+
+std::pair<rmm::device_vector<gdf_size_type>, rmm::device_vector<gdf_size_type>>
+generate_merged_indices(device_table const& left_table,
+                        device_table const& right_table,
+                        rmm::device_vector<int8_t> const& asc_desc,
+                        bool nulls_are_smallest,
+                        cudaStream_t stream) {
+
+    const gdf_size_type left_size  = left_table.num_rows();
+    const gdf_size_type right_size = right_table.num_rows();
+    const gdf_size_type total_size = left_size + right_size;
+
+    auto left_side = thrust::make_constant_iterator(static_cast<gdf_size_type>(LEFT_TABLE));
```

Comment:
I think this is a case where using the `make_*_iterator` is actually more verbose than just using the type directly:
```suggestion
    thrust::constant_iterator<gdf_size_type> left_side{LEFT_SIDE};;
```